WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -87,19 +87,41 @@ def index
     end
 
     if @display == 'admin'
-      @competitions = @competitions.includes(:delegates, :delegate_report)
+      @competitions = @competitions.includes(:delegate_report)
     end
 
     unless params[:region] == "all"
-      @competitions = @competitions.select { |competition| competition.belongs_to_region?(params[:region]) }
+      @competitions = @competitions.where("countryId = ? or continentId = ?", params[:region], params[:region])
     end
 
     if params[:search].present?
-      @competitions = @competitions.select { |competition| competition.contains?(params[:search]) }
+      params[:search].split.all? do |part|
+        @competitions = @competitions.where(
+          "Competitions.name like ? or
+          Competitions.cityName like ? or
+          users.name like ? or
+          Competitions.venue like ? or
+          Competitions.cellName like ? or
+          Competitions.countryId like ?",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+        )
+      end
     end
 
     unless params[:event_ids].empty?
-      @competitions = @competitions.select { |competition| competition.has_events_with_ids?(params[:event_ids]) }
+      params[:event_ids].each do |event_id|
+        ce_alias = "competition_events" + event_id
+        e_alias = "Events" + event_id
+        @competitions = @competitions.joins(
+          "join competition_events " + ce_alias + " ON " + ce_alias + ".competition_id = Competitions.id " \
+          "join Events " + e_alias + " ON " + e_alias + ".id = " + ce_alias + ".event_id",
+        ).where(e_alias + ".id = ?", event_id)
+      end
```

Comment:
I like the ideas here and they would certainly be beneficial not only for this PR but any future performance tuning. What is the best way to implement these?

I believe it's worth putting these change in before we implement any of the above though. I don't think there will be any improvements on the initial /competitions page but anytime an event is selected, or a search is entered, we'll now be executing 30 SQL statements as opposed to over 500. Below is my analysis.

**Main page: `/competitions`**
The current implementation runs a full table scan on Competitions and then uses
`  def country
    Country.c_find(countryId)
  end`
to get each Country.

The new implementation does a full table scan of Competitions and then eager loads the Countries.
Logs show 30 SQL queries for loading the page for both versions.

**With a search: `/competitions?utf8=✓&region=all&search=Results+Barry&year=all+years&state=past&display=list`**
The current implementation does a full table scan on Competitions and then runs an additional query against `users` and `competition_delegates` for each Competition returned. It then filters the results by the search terms in Ruby.

The new implementation does a full table scan on Competitions joined to `competition_delegates`, and `users`, filters the results in the database memory to the search term.
Logs show 560 SQL statements for the original and 31 for the new

**All events selected `/competitions?utf8=✓&event_ids%5B%5D=333&event_ids%5B%5D=444&event_ids%5B%5D=555&event_ids%5B%5D=222&event_ids%5B%5D=333bf&event_ids%5B%5D=333oh&event_ids%5B%5D=333fm&event_ids%5B%5D=333ft&event_ids%5B%5D=minx&event_ids%5B%5D=pyram&event_ids%5B%5D=sq1&event_ids%5B%5D=clock&event_ids%5B%5D=skewb&event_ids%5B%5D=666&event_ids%5B%5D=777&event_ids%5B%5D=444bf&event_ids%5B%5D=555bf&event_ids%5B%5D=333mbf&region=all&search=&year=all+years&state=past&display=list`**

The current implementation does a full scan on `Competitions` and then does an extra query to `Events` joined to `competition_events` for each row.

The new implementation does a full scan on `Competitions` which is joined to `competition_events` and `Events` once for each event selected (all in this case).

Logs show 558 SQL statements for the current version and 30 for the new version.